pkg/konfluxgen/konfluxgen.go

```diff
@@ -1192,3 +1210,110 @@ func consistentVersion(cfg Config, csv *operatorsv1alpha1.ClusterServiceVersion)
 	}
 	return version
 }
+
+type ReleaseConfig struct {
+	Snapshot            string
+	ReleasePlan         string
+	Environment         string
+	ResourcesOutputPath string
+}
+
+type Release struct {
+	Name        string
+	Snapshot    string
+	ReleasePlan string
+}
+
+func GenerateRelease(cfg ReleaseConfig) error {
+	fullOutputPath := filepath.Join(cfg.ResourcesOutputPath, ReleasesDirName)
+	// make sure output path exists
+	if err := os.MkdirAll(fullOutputPath, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to create directory %q: %w", fullOutputPath, err)
+	}
+
+	releaseName, err := unusedReleaseName(fullOutputPath, cfg.ReleasePlan)
+	if err != nil {
+		return fmt.Errorf("failed to generate release name for %q: %w", cfg.ReleasePlan, err)
+	}
+
+	releaseFileName := fmt.Sprintf("%s.yaml", cfg.ReleasePlan)
+
+	data := Release{
+		Name:        releaseName,
+		Snapshot:    cfg.Snapshot,
+		ReleasePlan: cfg.ReleasePlan,
+	}
+
+	if err := executeReleaseTemplate(data, filepath.Join(fullOutputPath, releaseFileName)); err != nil {
+		return fmt.Errorf("failed to execute release template: %w", err)
+	}
+
+	return nil
+}
+
+func executeReleaseTemplate(data Release, outputFilePath string) error {
+	funcs := template.FuncMap{
+		"sanitize": Sanitize,
+		"truncate": Truncate,
+	}
+
+	tpl, err := template.New("release.template.yaml").Delims("{{{", "}}}").Funcs(funcs).ParseFS(ReleaseTemplate, "*.yaml")
+	if err != nil {
+		return fmt.Errorf("failed to parse Release template: %w", err)
+	}
+
+	buf := &bytes.Buffer{}
+	if err := tpl.Execute(buf, data); err != nil {
+		return fmt.Errorf("failed to execute template for Release: %w", err)
+	}
+	if err := os.WriteFile(outputFilePath, buf.Bytes(), 0777); err != nil {
+		return fmt.Errorf("failed to write Release file %q: %w", outputFilePath, err)
+	}
+
+	return nil
+}
+
+func unusedReleaseName(outputPath, releasePlanName string) (string, error) {
+	releaseFilePath := filepath.Join(outputPath, fmt.Sprintf("%s.yaml", releasePlanName))
+	if !fileExists(releaseFilePath) {
+		return releasePlanName, nil
+	}
+
+	releaseMetadata, err := util.K8sMetadata(releaseFilePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read Release metadata: %w", err)
+	}
+
+	/*
+		release name is either:
+		<releasePlanName> or <releasePlanName>-<counter>
+	*/
+	if releaseMetadata.Name == releasePlanName {
+		return fmt.Sprintf("%s-2", releasePlanName), nil
+	} else if strings.HasPrefix(releaseMetadata.Name, fmt.Sprintf("%s-", releasePlanName)) {
+		strCounter := strings.TrimPrefix(releaseMetadata.Name, fmt.Sprintf("%s-", releasePlanName))
+		counter, err := strconv.Atoi(strCounter)
+		if err != nil {
+			return "", fmt.Errorf("could not parse suffix %q of %q to number", strCounter, releaseMetadata.Name)
+		}
+
+		return fmt.Sprintf("%s-%d", releasePlanName, counter+1), nil
+	} else {
+		return "", fmt.Errorf("last release name %q does not match pattern '%s-<counter>'", releaseMetadata.Name, releasePlanName)
+	}
+}
```

Comment:
this will always increment the counter even on the ones we don't really want/need to increase

Reply:
I think it's ok to not have support for retries and do it manually / document how to do it

Reply:
what we need to ensure is that this doesn't override existing ones if they have being modified for retrying